File: src/main/java/thecudster/sre/features/impl/dungeons/FragStatus.java
/*
 * SkyblockReinvented - Hypixel Skyblock Improvement Modification for Minecraft
 *  Copyright (C) 2021 theCudster
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as published
 *  by the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Affero General Public License for more details.
 *
 *  You should have received a copy of the GNU Affero General Public License
 *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 */

package thecudster.sre.features.impl.dungeons;

import com.google.gson.JsonArray;
import com.google.gson.JsonIOException;
import com.google.gson.JsonObject;
import thecudster.sre.util.api.APIUtil;

import java.net.MalformedURLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.stream.Collectors;

public class FragStatus {

  public static List<FragBot> getBots() throws JsonIOException {

    List<FragBot> bots = new ArrayList<>();

    JsonObject object = APIUtil.getJSONResponse("https://fragrunner.me:3001/fragbots");
    System.out.println(object);

    JsonArray jsonArray = object.getAsJsonArray("bots");
    for(int i = 0; i < jsonArray.size(); i++){
      JsonObject bot = jsonArray.get(i).getAsJsonObject();
      JsonArray array = bot.getAsJsonArray("queue");
      List<String> queue = new ArrayList<>();
      for(int a = 0; a < array.size(); a++) {
        queue.add(array.toString());
      }
      bots.add(
          new FragBot().setType("group").setStatus(bot.get("status").getAsString()).setUsername(bot.get("name").getAsString()).setUuid(bot.get("uuid").getAsString()).setQueue(queue)
      );
      System.out.println("Bots List: ");
      bots.forEach((botObj) -> System.out.println(botObj.getUsername()  + " : " + botObj.getType() + " : " + botObj.getStatus()));
    }
    return bots;
  }

  public static List<FragBot> getBestBot(int amount) throws MalformedURLException, JsonIOException {
    List<FragBot> bots = getBots();
    bots = bots.stream().filter(bot -> bot.getStatus().equals("online")).collect(Collectors.toList());
    if(bots.size() == 0) return new ArrayList<>();
    if(amount > bots.size()){
      amount = bots.size();
    }
    List<FragBot> out = new ArrayList<>();
    for(int i = 0; i < amount; i++){
      int random = new Random().nextInt(bots.size());
      FragBot bot = bots.get(random);
      if(out.contains(bot)){
        i--;
        continue;
      }
      out.add(bot);
    }
    return out;
  }
}

File: src/main/kotlin/thecudster/sre/features/impl/dungeons/FragStatus.kt
/*
 * SkyblockReinvented - Hypixel Skyblock Improvement Modification for Minecraft
 *  Copyright (C) 2021 theCudster
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as published
 *  by the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Affero General Public License for more details.
 *
 *  You should have received a copy of the GNU Affero General Public License
 *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 */
package thecudster.sre.features.impl.dungeons

import com.google.gson.JsonIOException
import thecudster.sre.util.api.APIUtil.getJSONResponse
import java.util.function.Consumer
import java.net.MalformedURLException
import thecudster.sre.features.impl.dungeons.FragStatus
import java.util.*
import java.util.stream.Collectors

object FragStatus {
    @get:Throws(JsonIOException::class)
    val bots: List<FragBot>
        get() {
            val bots: MutableList<FragBot> = ArrayList()
            val `object` = getJSONResponse("https://fragrunner.me:3001/fragbots")
            println(`object`)
            val jsonArray = `object`.getAsJsonArray("bots")
            for (i in 0 until jsonArray.size()) {
                val bot = jsonArray[i].asJsonObject
                val array = bot.getAsJsonArray("queue")
                val queue: MutableList<String> = ArrayList()
                for (a in 0 until array.size()) {
                    queue.add(array.toString())
                }
                bots.add(
                    FragBot().setType("group").setStatus(bot["status"].asString).setUsername(bot["name"].asString)
                        .setUuid(bot["uuid"].asString).setQueue(queue)
                )
                println("Bots List: ")
                bots.forEach(Consumer { botObj: FragBot -> println(botObj.username + " : " + botObj.type + " : " + botObj.status) })
            }
            return bots
        }

    @JvmStatic
    @Throws(MalformedURLException::class, JsonIOException::class)
    fun getBestBot(amount: Int): List<FragBot> {
        var amount = amount
        var bots = bots
        bots = bots.stream().filter { bot: FragBot -> bot.status == "online" }.collect(Collectors.toList())
        if (bots.size == 0) return ArrayList()
        if (amount > bots.size) {
            amount = bots.size
        }
        val out: MutableList<FragBot> = ArrayList()
        var i = 0
        while (i < amount) {
            val random = Random().nextInt(bots.size)
            val bot = bots[random]
            if (out.contains(bot)) {
                i--
                i++
                continue
            }
            out.add(bot)
            i++
        }
        return out
    }
}